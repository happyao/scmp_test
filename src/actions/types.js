export const types = {
  GET_DATA: 'GET_DATA',
  CHOOSE_ITEM: 'CHOOSE_ITEM',
  VOTE_ITEM: 'VOTE_ITEM'
}
