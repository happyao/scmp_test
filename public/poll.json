{
  "polls": [
    {
      "id": 1,
      "title": "Is bitcoin worth the time and money that mining requires?",
      "publishedDate": 1516605447,
      "answer": {
        "type": "Single",
        "options": [
          {
            "id": 1,
            "label": "Yes",
            "vote": 2
          },
          {
            "id": 2,
            "label": "No",
            "vote": 2
          }
        ]
      }
    },
    {
      "id": 2,
      "title": "Should chatbots replace humans in customer service jobs?",
      "publishedDate": 1516000647,
      "answer": {
        "type": "Single",
        "options": [
          {
            "id": 3,
            "label": "Yes",
            "vote": 4
          },
          {
            "id": 4,
            "label": "No",
            "vote": 2
          }
        ]
      }
    },
    {
      "id": 3,
      "title": "How are we feeling about 2018?",
      "publishedDate": 1515568647,
      "answer": {
        "type": "Single",
        "options": [
          {
            "id": 5,
            "label": "Hopeful",
            "vote": 2
          },
          {
            "id": 6,
            "label": "Doubtful",
            "vote": 2
          }
        ]
      }
    },
    {
      "id": 4,
      "title": "Which country/region have you ever visited? (Select all that applies)",
      "publishedDate": 1515482247,
      "answer": {
        "type": "Multi",
        "options": [
          {
            "id": 7,
            "label": "Hong Kong",
            "vote": 2
          },
          {
            "id": 8,
            "label": "China",
            "vote": 1
          },
          {
            "id": 9,
            "label": "Australia",
            "vote": 1
          },
          {
            "id": 10,
            "label": "Thailand",
            "vote": 1
          },
          {
            "id": 11,
            "label": "Korea",
            "vote": 1
          },
          {
            "id": 12,
            "label": "Japan",
            "vote": 3
          }
        ]
      }
    },
    {
      "id": 5,
      "title": "Will new benefits encourage you to study or work in mainland?",
      "publishedDate": 1515309447,
      "answer": {
        "type": "Single",
        "options": [
          {
            "id": 13,
            "label": "Yes",
            "vote": 2
          },
          {
            "id": 14,
            "label": "No",
            "vote": 1
          }
        ]
      }
    },
    {
      "id": 6,
      "title": "hi, hi another question 2?",
      "publishedDate": 1515309447,
      "answer": {
        "type": "Single",
        "options": [
          {
            "id": 14,
            "label": "Yes",
            "vote": 2
          },
          {
            "id": 15,
            "label": "No",
            "vote": 1
          }
        ]
      }
    },
    {
      "id": 7,
      "title": "hi another question?",
      "publishedDate": 1515309447,
      "answer": {
        "type": "Single",
        "options": [
          {
            "id": 16,
            "label": "Yes",
            "vote": 2
          },
          {
            "id": 17,
            "label": "No",
            "vote": 1
          }
        ]
      }
    },
    {
      "id": 8,
      "title": "hi another question 3?",
      "publishedDate": 1515309447,
      "answer": {
        "type": "Single",
        "options": [
          {
            "id": 18,
            "label": "Yes",
            "vote": 2
          },
          {
            "id": 19,
            "label": "No",
            "vote": 1
          }
        ]
      }
    },
    {
      "id": 9,
      "title": "hi another question 4?",
      "publishedDate": 1515309447,
      "answer": {
        "type": "Single",
        "options": [
          {
            "id": 20,
            "label": "Yes",
            "vote": 2
          },
          {
            "id": 21,
            "label": "No",
            "vote": 1
          }
        ]
      }
    }
  ]
}
