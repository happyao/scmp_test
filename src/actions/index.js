import { types } from './types'

export const chooseItem = id => ({
  type: types.CHOOSE_ITEM,
  id
})

export const voteItem = id => ({
  type: types.VOTE_ITEM,
  id
})
