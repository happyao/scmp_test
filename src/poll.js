var customData = {
  polls: [
    {
      id: 1,
      title: 'Is bitcoin worth the time and money that mining requires?',
      publishedDate: 1516605447,
      answer: {
        type: 'Single',
        options: [
          {
            id: 1,
            label: 'Yes',
            vote: 10
          },
          {
            id: 2,
            label: 'No',
            vote: 20
          }
        ]
      }
    },
    {
      id: 2,
      title: 'Should chatbots replace humans in customer service jobs?',
      publishedDate: 1516000647,
      answer: {
        type: 'Single',
        options: [
          {
            id: 3,
            label: 'Yes',
            vote: 30
          },
          {
            id: 4,
            label: 'No',
            vote: 40
          }
        ]
      }
    },
    {
      id: 3,
      title: 'How are we feeling about 2018?',
      publishedDate: 1515568647,
      answer: {
        type: 'Single',
        options: [
          {
            id: 5,
            label: 'Hopeful',
            vote: 50
          },
          {
            id: 6,
            label: 'Doubtful',
            vote: 60
          }
        ]
      }
    },
    {
      id: 4,
      title:
        'Which country/region have you ever visited? (Select all that applies)',
      publishedDate: 1515482247,
      answer: {
        type: 'Multi',
        options: [
          {
            id: 7,
            label: 'Hong Kong',
            vote: 70
          },
          {
            id: 8,
            label: 'China',
            vote: 80
          },
          {
            id: 9,
            label: 'Australia',
            vote: 90
          },
          {
            id: 10,
            label: 'Thailand',
            vote: 100
          },
          {
            id: 11,
            label: 'Korea',
            vote: 110
          },
          {
            id: 12,
            label: 'Japan',
            vote: 120
          }
        ]
      }
    },
    {
      id: 5,
      title: 'Will new benefits encourage you to study or work in mainland?',
      publishedDate: 1515309447,
      answer: {
        type: 'Single',
        options: [
          {
            id: 13,
            label: 'Yes',
            vote: 120
          },
          {
            id: 14,
            label: 'No',
            vote: 100
          }
        ]
      }
    }
  ]
}

module.exports = customData
