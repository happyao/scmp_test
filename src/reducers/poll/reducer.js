import { types } from '../../actions/types'
// import customData from '../../poll'

export const initialState = {
  currentID: 0,
  customData: []
}
export default (state = initialState, action) => {
  switch (action.type) {
    case types.GET_DATA:
      return { ...state, customData: action.customData }
    case types.CHOOSE_ITEM:
      return { ...state, currentID: action.id }
    case types.VOTE_ITEM:
      return {
        ...state,
        customData: state.customData.map(cd => {
          if (state.currentID === cd.id - 1) {
            return {
              ...cd,
              answer: {
                ...cd.answer,
                options: cd.answer.options.map(item => {
                  if (item.id === action.id) {
                    return { ...item, vote: item.vote + 1 }
                  } else {
                    return item
                  }
                })
              }
            }
          } else {
            return cd
          }
        })
      }

    default:
      return state
  }
}
