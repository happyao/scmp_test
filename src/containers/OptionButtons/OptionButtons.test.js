import React from 'react'
import { OptionButtons } from './index'
import poll from '../../poll'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'
import * as actions from '../../actions'
describe('OptionButtons', () => {
  it('renders without crashing given the required props', () => {
    const props = {
      currentID: 2,
      answers: poll.polls[2].answer,
      colors: ['#143b6b', '#e57335']
    }

    const wrapper = shallow(<OptionButtons {...props} />)
    expect(toJson(wrapper)).toMatchSnapshot()
  })
  it('renders right component given the required props', () => {
    const props = {
      currentID: 3,
      answers: poll.polls[3].answer,
      colors: ['#143b6b', '#e57335']
    }
    const wrapper = shallow(<OptionButtons {...props} />)
    expect(wrapper.find('button').length).toBe(6)
  })

  it('inactive buttons when click single type poll option', () => {
    const props = {
      currentID: 2,
      answers: poll.polls[2].answer,
      colors: ['#143b6b', '#e57335'],
      voteItem: jest.fn()
    }

    const state = {
      disabled: true,
      historicalId: [2]
    }

    actions.voteItem = jest.fn()
    const wrapper = shallow(<OptionButtons {...props} />)

    wrapper
      .find('button')
      .at(0)
      .simulate('click', { target: { style: { backgroundColor: '#fffff' } } })
    expect(wrapper.state()).toStrictEqual(state)
    expect(
      wrapper
        .find('button')
        .at(0)
        .props()['disabled']
    ).toBe(true)
  })

  it('inactive buttons when click multi type poll option', () => {
    const props = {
      currentID: 3,
      answers: poll.polls[3].answer,
      colors: ['#143b6b', '#e57335'],
      voteItem: jest.fn()
    }

    const state = {
      disabled: false,
      historicalId: [3]
    }

    actions.voteItem = jest.fn()
    const wrapper = shallow(<OptionButtons {...props} />)

    wrapper
      .find('button')
      .at(0)
      .props()
      .onClick({
        target: { style: { backgroundColor: '#fffff' } }
      })
    expect(props.voteItem.mock.calls.length).toBe(1)
    expect(wrapper.state()).toStrictEqual(state)
  })
})
