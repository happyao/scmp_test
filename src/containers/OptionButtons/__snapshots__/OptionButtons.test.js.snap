// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OptionButtons inactive buttons when click single type poll option 1`] = `
<Fragment>
  <div
    className="options"
  >
    <button
      disabled={true}
      key="5"
      onClick={[Function]}
      style={
        Object {
          "backgroundColor": "#143b6b",
        }
      }
    >
      Hopeful
    </button>
    <button
      disabled={true}
      key="6"
      onClick={[Function]}
      style={
        Object {
          "backgroundColor": "#e57335",
        }
      }
    >
      Doubtful
    </button>
  </div>
  <div
    style={
      Object {
        "color": "#143b6b",
        "fontSize": "1.5rem",
        "fontWeight": "bold",
        "visibility": "visible",
      }
    }
  >
    You have submitted your answer.
  </div>
</Fragment>
`;

exports[`OptionButtons renders without crashing given the required props 1`] = `
<Fragment>
  <div
    className="options"
  >
    <button
      disabled={false}
      key="5"
      onClick={[Function]}
      style={
        Object {
          "backgroundColor": "#143b6b",
        }
      }
    >
      Hopeful
    </button>
    <button
      disabled={false}
      key="6"
      onClick={[Function]}
      style={
        Object {
          "backgroundColor": "#e57335",
        }
      }
    >
      Doubtful
    </button>
  </div>
  <div
    style={
      Object {
        "color": "#143b6b",
        "fontSize": "1.5rem",
        "fontWeight": "bold",
        "visibility": "hidden",
      }
    }
  >
    You have submitted your answer.
  </div>
</Fragment>
`;
